### PR TITLE
perf(mtp): Gemma4 MTP sliding-window fix + 8-bit benchmark results

### PR DESCRIPTION
**Pull request by solderzzc**
## Summary

Resolves throughput regression in Gemma4 MTP speculative decoding at long-context lengths, and documents precision-dependent MTP behaviour from empirical benchmarking.

### mlx-swift-lm submodule (→ fix/compiler-warnings-mtp-optim @ c552b4d)
- **Sliding-window KV cap**: runMTPHead caps shared-KV cross-attention to last 16 backbone positions (O(T) → O(16)). Eliminates 2–4× throughput regression at 40K–100K context.
- **MTPPartialRollback protocol**: stores lastBackboneHiddenStateAll for partial-rejection rollback without re-running the main model.
- **callMTPHeadOnly**: re-seeds MTP head from cached backbone state at near-zero cost.
- numMTPDraftTokens=2 (depth=4 empirically slower on Metal).

### Benchmark Results (M5 Pro 64 GB, gemma-4-26b-a4b-it-8bit)

8-bit is bandwidth-bound → KV reads amortize across the batch → MTP provides real gains:

| Config | 512 ctx | 40K ctx | 100K ctx |
|--------|---------|---------|---------|
| Vanilla | 53.7 | 32.4 | 14.9 |
| **Vanilla + MTP** ⭐ | 47.1 | **38.8 (+20%)** | **22.5 (+51%)** |
| Vanilla + TurboQuant | 53.5 | **50.1** | **48.3** |
| TQ + MTP | 47.4 | 31.0 | 23.3 |

4-bit MoE is compute-bound; MTP neutral. TQ+MTP counterproductive at both precisions.

### Other
- mtp_bench.py: Metal warmup request before first timed run (fixes inflated 1.77s TTFT)
- README: split 4-bit/8-bit tables, precision-specific guidance
- Server/CLI: --mtp / --num-mtp-tokens wired through GenerationConfig

## User Guidance

| Precision | Best long-context config |
|-----------|--------------------------|
| **8-bit** | --mtp alone (+20–51%) |
| **8-bit, memory-critical** | --turbo-kv alone |
| **4-bit MoE** | --turbo-kv alone |
| **4-bit + both** | Skip MTP |

### CODE REVIEW

**Comment by solderzzc** [comment]
Reviewed against current main. The textual conflict is small — one submodule pointer — but there is a semantic collision that needs a decision before any of this lands, so recording the analysis rather than merging.

## The decision

Main grew its own MTP path while this sat: `--mtp` / `--num-mtp-tokens` exist (`Server.swift:283-287`) and route through `MLXLMCommon.generateMTP(…)` (`Server.swift:1654`, `:1665`), guarded on `config.mtp` and `context.model is any MTPLanguageModel`.

This PR adds a **competing** branch (`else if let asstRef = mtpAsstRef`) placed *before* the prompt-cache branches. After a clean auto-merge you get both, and this one wins whenever an assistant checkpoint is in the HF cache — silently bypassing the native path *and* the prompt cache. Nothing errors; it just quietly takes a different route.

**So: native `generateMTP`, or the dual-model assistant path?** That is yours to call, and it decides whether the server wiring here is wanted at all.

## What is genuinely unlanded and worth having

- **`maxSharedKV=16` in `runMTPHead`** (O(T) → O(16) cross-attention). No `maxSharedKV` / `MTPPartialRollback` / `callMTPHeadOnly` anywhere in main's submodule. This fixes a 2-4× regression at 40K-100K context and merges cleanly — the submodule hunks are disjoint from main's Gemma 4 KV-shared work (main touches lines 47-1010, this touches 990+).
- **The Test 13 harness fix.** `run_benchmark.sh:1360` still builds `--product Gemma4MTPBench`, which **is not a product in `Package.swift`** — Test 13 on main cannot run at all. The `scripts/profiling/mtp_bench.py` replacement works. This is standalone and un-breaks something currently broken.
- The `MTP_ACCEPTED` acceptance-rate logging is a clean freebie either way.

## Suggested split

1. **Land now:** the Test 13 harness fix. No conflicts, repairs a broken benchmark.
2. **Land via an `mlx-swift-lm` PR:** the `maxSharedKV` cap. Its branch (`c552b4d`) is 6 commits behind main's submodule — it predates the weight-index guard (`e60ccfd`) and the Gemma 4 KV-shared loader (`b320bc4`) — so it needs one bench run against the current loader to confirm the win holds.
3. **Decide, then keep or drop:** the `mtpAsstRef` / `Gemma4MTPRegistry` server wiring. Note the registry's hardcoded table maps `gemma-4-e4b-it` → the *E2B* assistant and `gemma-4-31b-it` → the *26B* assistant, which look like transcription slips.
4. **Reconcile by hand:** the README tables. Main has a Gemma-4-26B MTP table at README:55-90 (71.5 / 65.2 tok/s); this rewrites a different section with a different run of the same model (77.5 / 66.9). Git merges both, and you would ship two contradictory tables.

Also: the branch adds `test_array_init.swift` and `test_scatter.swift` at the submodule repo root; those want dropping.

**Comment by solderzzc** [comment]
Followed up on the "native `generateMTP` vs dual-model" question I posed earlier. Having read the code, **that framing was wrong — they are not competing implementations of the same thing.** They serve different model families, and the native path structurally cannot serve Gemma 4.

## How MTP is actually structured

`MTPLanguageModel` (`MLXLMCommon/LanguageModel.swift:255`) expects the MTP heads to live *inside* the model: `callMTP` returns `[main_logits, mtp_0_logits, …]` from one pass over one checkpoint. Conformances in main:

- `Qwen35Model`, `Qwen35TextModel` (`Qwen35.swift:896`, `:948`)
- `DeepseekV4Model` (`DeepseekV4.swift:878`)

Those checkpoints carry `mtp.*` weights in-file — the same tensors that turned up in #118, incidentally.

**Gemma 4 has no in-model MTP heads.** `Gemma4Model` (`Gemma4.swift:57`) and `Gemma4TextModel` (`Gemma4Text.swift:947`) conform to `LLMModel` only. Google ships MTP as a *separate assistant checkpoint*, which is why `DualModelMTP` (`LanguageModel.swift:276`) exists at all — it is `MTPLanguageModel` plus a `mainModelRef` back-pointer, and `Gemma4AssistantModel` (`Gemma4Text.swift:1079`) conforms to it.

## What that means for main today

`Server.swift:1629` and `:1640` gate on `config.mtp, context.model is any MTPLanguageModel`. For any Gemma 4 model that test is **false**, so `--mtp` is silently a no-op there — no error, no log line, nothing.

Meanwhile the dual-model half is present but unreachable: nothing in `Sources/` sets `mainModelRef` except `Sources/Gemma4MTPBench/main.swift:173`, and **`Gemma4MTPBench` is not a target in `Package.swift`** — that directory cannot build. So the assistant path is dead code in main, and `run_benchmark.sh:1360` builds a product that does not exist, which is why Test 13 cannot run.

So this PR is not a competing implementation. It is the missing wiring for a family the native path cannot cover.

## What I would change about it

The objections from my earlier comment stand, but they are placement and hygiene, not architecture:

1. **Branch placement.** Putting `else if let asstRef = mtpAsstRef` before the prompt-cache branches means it bypasses the prompt cache. It belongs alongside the existing `config.mtp` test.
2. **Better: no second branch at all.** `generateMTP` already accepts any `MTPLanguageModel`, and `DualModelMTP` *is* one. If the server sets `mainModelRef` when the loaded model conforms to `DualModelMTP`, the existing call site handles both families unchanged, and the difference stays a conformance detail rather than a fork in the server:

   ```swift
   if config.mtp, let dual = context.model as? any DualModelMTP {
       dual.mainModelRef = mainContext.model   // assistant needs the trunk
   }
   if config.mtp, context.model is any MTPLanguageModel {
       stream = try MLXLMCommon.generateMTP(...)   // unchanged
   }
   ```
   That keeps one path, keeps the prompt cache, and drops the parallel branch entirely.
3. **The registry.** `Gemma4MTPRegistry`'s hardcoded table maps `gemma-4-e4b-it` → the *E2B* assistant and `gemma-4-31b-it` → the *26B* assistant, which look like transcription slips. Worth replacing with an explicit `--mtp-assistant-model` flag, or config-driven detection, rather than a guessed table.

## Suggested order, unchanged otherwise

1. **Test 13 fix** — standalone, un-breaks a benchmark that currently cannot build. Note it will also need `Gemma4MTPBench` added to `Package.swift`, or the `mtp_bench.py` replacement (which avoids the target entirely — probably the better answer).
2. **`maxSharedKV=16`** via an `mlx-swift-lm` PR — genuinely unlanded, fixes a 2–4× long-context regression, merges cleanly.
3. **The wiring**, reshaped as (2) above.
4. **README tables** reconciled by hand.

Happy to take any of these if useful.

**Comment by solderzzc** [comment]
Closing as superseded. This PR bundled three things, and each has been resolved independently:

**MTP-assistant wiring** — superseded by #137 (merged), which ships the same capability more generally: an explicit `--mtp-assistant-model` flag rather than this PR's hardcoded Gemma4-name-prefix auto-resolution (`Gemma4MTPRegistry`). Rebasing this PR forward would have landed a second, narrower implementation of a feature that already works for any `DualModelMTP` architecture, not just Gemma4. Investigated in detail before making that call — the two wiring paths touch the same functions in incompatible ways, so this wasn't a rebase-vs-don't-rebase judgment call, it was a genuine conflict between two designs.

**Sliding-window KV cap** — independently ported and evaluated on this session's hardware. No measurable benefit at either 150 tokens (62.6 vs 63.6 tok/s) or ~9k context (18.5 vs 18.6 tok/s) on the 4-bit model available for testing. Parked rather than merged, since the numbers didn't support it in the regime I could reach. The 8-bit long-context claim in this PR's benchmark tables (2–4× at 40K–100K) may still hold — I couldn't test 8-bit at that context length — so this isn't a refutation, just an unreproduced-in-this-environment result. If anyone revisits it, the ported branch is `perf/mtp-shared-kv-window` locally; start there rather than from this PR's submodule pointer, which now predates the dual-model MTP crash fix (mlx-swift-lm#46) by a wide margin.

**Benchmark tooling** — landed via #151. `mtp_bench.py`, the `run_benchmark.sh` Test 13 fix (the old invocation called a `Gemma4MTPBench` product that no longer exists in `Package.swift` — that path was already broken on main independent of this PR), and the README's benchmark tables.

Thanks for the original investigation — the 8-bit MTP numbers here (`+20%` at 40K, `+51%` at 100K) were genuinely useful as reference data even where the KV-cap mechanism itself didn't reproduce elsewhere.
